File: Problems/7_transform_matrix/solution.py
import numpy as np

import numpy as np

def transform_matrix(A: list[list[int|float]], T: list[list[int|float]], S: list[list[int|float]]) -> list[list[int|float]]:
	#Define determinant of a matrix to check if invertible
	if not np.linalg.det(T) or not np.linalg.det(S):
		return -1
	#Invert T
	TInv = np.linalg.inv(T)
	#Multiply to check if it works
	return np.dot(TInv,A).dot(S)

def test_transform_matrix() -> None:
    # Test cases for transform_matrix function
    # Test case 1
    A = [[1, 2], [3, 4]]
    T = [[2, 0], [0, 2]]
    S = [[1, 1], [0, 1]]
    assert transform_matrix(A, T, S) == [[0.5, 1.5], [1.5, 3.5]]
    
    # Test case 2
    A = [[1, 0], [0, 1]]
    T = [[1, 2], [3, 4]]
    S = [[2, 0], [0, 2]]
    assert transform_matrix(A, T, S) == [[-4.0, 2.0], [3.0, -1.0]]
    
    # Test case 3
    A = [[2, 3], [1, 4]]
    T = [[3, 0], [0, 3]]
    S = [[1, 1], [0, 1]]
    assert transform_matrix(A, T, S) == [[0.667, 1.667], [0.333, 1.667]]
    
    # Test case 4
    A = [[2, 3], [1, 4]]
    T = [[3, 0], [0, 3]]
    S = [[1, 1], [1, 1]]
    assert transform_matrix(A, T, S) == -1
    # try:
    #     transform_matrix(A, T, S)
    #     assert False, "ValueError not raised"
    # except ValueError as e:
    #     assert str(e) == "The matrices T and/or S are not invertible."

if __name__ == "__main__":
    test_transform_matrix()
    print("All transform_matrix tests passed.")

File: Problems/7_transform_matrix/solution.cpp
#include <iostream>
#include <eigen3/Eigen/Dense>
#include <cmath>

using namespace Eigen;
using namespace std;

// Transform matrix A using transformation matrices T and S.
// If either T or S is non-invertible, return a 1x1 matrix with the single element -1.
MatrixXd transform_matrix(const MatrixXd &A, const MatrixXd &T, const MatrixXd &S) {
    // Check if T and S are invertible (using a small threshold to account for floating-point precision)
    if (fabs(T.determinant()) < 1e-9 || fabs(S.determinant()) < 1e-9) {
        MatrixXd error(1, 1);
        error(0, 0) = -1;
        return error;
    }

    // Compute the inverse of T
    MatrixXd T_inv = T.inverse();

    // Perform the matrix transformation: T_inv * A * S
    MatrixXd result = T_inv * A * S;

    return result;
}

void test_transform_matrix() {
    {
        // Test case 1
        MatrixXd A(2, 2), T(2, 2), S(2, 2);
        A << 1, 2,
             3, 4;
        T << 2, 0,
             0, 2;
        S << 1, 1,
             0, 1;
        MatrixXd result = transform_matrix(A, T, S);
        MatrixXd expected(2, 2);
        expected << 0.5, 1.5,
                    1.5, 3.5;
        if (!result.isApprox(expected, 1e-3)) {
            cout << "Test case 1 failed.\n";
            cout << "Result:\n" << result << "\nExpected:\n" << expected << "\n";
            exit(1);
        }
    }

    {
        // Test case 2
        MatrixXd A(2, 2), T(2, 2), S(2, 2);
        A << 1, 0,
             0, 1;
        T << 1, 2,
             3, 4;
        S << 2, 0,
             0, 2;
        MatrixXd result = transform_matrix(A, T, S);
        MatrixXd expected(2, 2);
        expected << -4.0, 2.0,
                     3.0, -1.0;
        if (!result.isApprox(expected, 1e-3)) {
            cout << "Test case 2 failed.\n";
            cout << "Result:\n" << result << "\nExpected:\n" << expected << "\n";
            exit(1);
        }
    }

    {
        // Test case 3
        MatrixXd A(2, 2), T(2, 2), S(2, 2);
        A << 2, 3,
             1, 4;
        T << 3, 0,
             0, 3;
        S << 1, 1,
             0, 1;
        MatrixXd result = transform_matrix(A, T, S);
        MatrixXd expected(2, 2);
        expected << 0.667, 1.667,
                    0.333, 1.667;
        if (!result.isApprox(expected, 1e-3)) {
            cout << "Test case 3 failed.\n";
            cout << "Result:\n" << result << "\nExpected:\n" << expected << "\n";
            exit(1);
        }
    }

    {
        // Test case 4: S is non-invertible; expected result is a 1x1 matrix with -1.
        MatrixXd A(2, 2), T(2, 2), S(2, 2);
        A << 2, 3,
             1, 4;
        T << 3, 0,
             0, 3;
        S << 1, 1,
             1, 1;  // This matrix is singular.
        MatrixXd result = transform_matrix(A, T, S);
        if (result.rows() != 1 || result.cols() != 1 || result(0, 0) != -1) {
            cout << "Test case 4 failed.\n";
            cout << "Result:\n" << result << "\nExpected: [-1]\n";
            exit(1);
        }
    }
}

int main() {
    test_transform_matrix();
    cout << "All transform_matrix tests passed.\n";
    return 0;
}